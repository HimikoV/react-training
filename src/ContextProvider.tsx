import { createContext, useEffect, useState, useMemo } from "react";
import { mapBasketData } from "./mappers/mapBasketData";
import { getDeliveryOptions } from "./service/getDeliveryOptions";
import { getOrderData } from "./service/getOrderData";
import { getPaymentOptions } from "./service/getPaymentOptions";
import { requestData } from "./service/requestData";
import { basketType } from "./types/basketType";
import { finalizeOptionsType } from "./types/finalizeOptionsType";
import { orderType } from "./types/orderType";
import { productType } from "./types/productType";
import { childrenInterface } from "./interfaces/childrenInterface";
import { basketProcessedType } from "./types/basketProcessedType";
import { contextProviderInterface } from "./interfaces/contextProviderInterface";

export const dataContext = createContext<contextProviderInterface>(
  {} as contextProviderInterface
);

const ContextProvider = ({ children }: childrenInterface) => {
  const [basketData, setBasketData] = useState<basketType[]>(
    JSON.parse(localStorage.getItem("basketData") || "[]")
  );
  const [filter, setFilter] = useState<string[]>([]);
  const [data, setData] = useState<productType[]>([]);
  const [loadLimit, setLoadLimit] = useState<number>(9);
  const [basketProcessedData, setBasketProcessedData] = useState<
    basketProcessedType[]
  >([]);
  const [deliveryOptions, setDeliveryOptions] = useState<finalizeOptionsType[]>(
    []
  );
  const [paymentOptions, setPaymentOptions] = useState<finalizeOptionsType[]>(
    []
  );
  const [open, setOpen] = useState<boolean>(false);
  const [orderData, setOrderData] = useState<orderType[]>([]);

  useEffect(() => {
    const fetchData = async () => {
      const responseData = await requestData();
      setData(responseData);
      const responseDelivery = await getDeliveryOptions();
      setDeliveryOptions(responseDelivery);
      const responsePayment = await getPaymentOptions();
      setPaymentOptions(responsePayment);
      const responseOrderData = await getOrderData();
      setOrderData(responseOrderData);
    };
    fetchData();
  }, []);

  useEffect(() => {
    localStorage.setItem("basketData", JSON.stringify(basketData));
  }, [basketData]);

  useEffect(() => {
    mapBasketData(data, basketData, setBasketProcessedData);
  }, [data]);

  const trueValues = useMemo(
    () => ({
      filter,
      setFilter,
      basketData,
      setBasketData,
      data,
      loadLimit,
      setLoadLimit,
      basketProcessedData,
      setBasketProcessedData,
      deliveryOptions,
      setDeliveryOptions,
      paymentOptions,
      setPaymentOptions,
      open,
      setOpen,
      orderData,
      setOrderData,
    }),
    [
      filter,
      setFilter,
      basketData,
      setBasketData,
      data,
      loadLimit,
      setLoadLimit,
      basketProcessedData,
      setBasketProcessedData,
      deliveryOptions,
      setDeliveryOptions,
      paymentOptions,
      setPaymentOptions,
      open,
      setOpen,
      orderData,
      setOrderData,
    ]
  );

  // const values = {};
  return (
    <dataContext.Provider value={trueValues}>{children}</dataContext.Provider>
  );
};
export default ContextProvider;
