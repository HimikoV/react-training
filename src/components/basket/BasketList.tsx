import { Grid } from "@mui/material";
import { useContext, useEffect } from "react";
import { dataContext } from "../../ContextProvider";
import BasketItem from "./BasketItem";
import { mapBasketData } from "../../mappers/mapBasketData";
import { basketProcessedType } from "../../types/basketProcessedType";

const BasketList = () => {
  const { basketData, data, basketProcessedData, setBasketProcessedData } =
    useContext(dataContext);

  useEffect(() => {
    mapBasketData(data, basketData, setBasketProcessedData);
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [data, basketData]);

  return (
    <Grid>
      {basketProcessedData.map((item: basketProcessedType) => {
        return <BasketItem key={item.key} item={item} />;
      })}
    </Grid>
  );
};
export default BasketList;
