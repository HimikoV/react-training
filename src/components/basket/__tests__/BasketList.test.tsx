import { describe, it } from "@jest/globals";
import renderer from "react-test-renderer";
import ContextProvider from "../../../ContextProvider";
import BasketList from "../BasketList";

describe("testing BasketList component", () => {
  it("BasketList Renders Correctly", () => {
    const basketList = renderer
      .create(
        <ContextProvider>
          <BasketList />
        </ContextProvider>
      )
      .toJSON();
    expect(basketList).toMatchSnapshot();
  });
});
