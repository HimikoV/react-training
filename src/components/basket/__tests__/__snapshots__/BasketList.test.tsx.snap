// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing BasketList component BasketList Renders Correctly 1`] = `
<div
  className="MuiGrid-root css-vj1n65-MuiGrid-root"
/>
`;
